internal/services/aws_iam_service.go

```diff
@@ -0,0 +1,58 @@
+package services
+
+import (
+	"net/http"
+
+	"github.com/RHEnVision/provisioning-backend/internal/clients"
+	"github.com/RHEnVision/provisioning-backend/internal/ctxval"
+	"github.com/RHEnVision/provisioning-backend/internal/models"
+	"github.com/RHEnVision/provisioning-backend/internal/payloads"
+	"github.com/go-chi/chi/v5"
+	"github.com/go-chi/render"
+)
+
+func ValidatePermissions(w http.ResponseWriter, r *http.Request) {
+	logger := ctxval.Logger(r.Context())
+	sourceId := chi.URLParam(r, "ID")
+	region := r.URL.Query().Get("region")
+	if region == "" {
+		renderError(w, r, payloads.NewMissingRequestParameterError(r.Context(), "region parameter is missing"))
+	}
+
+	// Get Sources client
+	sourcesClient, err := clients.GetSourcesClient(r.Context())
+	if err != nil {
+		renderError(w, r, payloads.NewClientError(r.Context(), err))
+		return
+	}
+
+	// Fetch arn from Sources
+	authentication, err := sourcesClient.GetAuthentication(r.Context(), sourceId)
+	if err != nil {
+		renderError(w, r, payloads.NewClientError(r.Context(), err))
+		return
+	}
+
+	if typeErr := authentication.MustBe(models.ProviderTypeAWS); typeErr != nil {
+		renderError(w, r, payloads.NewClientError(r.Context(), typeErr))
+		return
+	}
+
+	ec2Client, err := clients.GetEC2Client(r.Context(), authentication, region)
+	if err != nil {
+		renderError(w, r, payloads.NewAWSError(r.Context(), "unable to get AWS EC2 client", err))
+		return
+	}
+
+	logger.Info().Msgf("Listing policies attached to the role.")
+	policies, err := ec2Client.ListAttachedRolePolicies(r.Context(), authentication)
+	if err != nil {
+		renderError(w, r, payloads.NewAWSError(r.Context(), "unable to list AWS EC2 instances", err))
+		return
+	}
```

Comment:
Consider moving this code logic from here (service package) to the EC2 client (clients package) if you want to avoid creating too many custom types.

Or if you like it this way, I am okay, it is just more work. It is better because then we can re-use these functions later on. That is fine too, just little bit more work.

Reply:
Is this endpoint going to serve validation to Azure and GCP as well?

Reply:
That is my thinking going forward, yes. One of the reasons why I would prefer a simple string-based interface.